File: components/workout_txt_reader.py
import re
from components.helpers import *
from components.workout_data_structs import *

# regular expression describing a 'dd.mm' date format
date_pattern = r'\d{2}\.\d{2}'

# This list will contain lists of exercises that each represent an entire workout
workout_list_pre = []
# This list contains the same data as 'workout_list_pre' but as Workout objects
workout_list = []

# Open .txt containing workout data
def read_and_clean_workout_data(file_path):
    with open(file_path, 'r') as file:
        data = file.read()

    workouts = data.split('\n\n')

    # clear out single lines of text
    for workout in workouts:
        if '\n' not in workout:
            workouts.remove(workout)
    
    return workouts

# Take the data read from the .txt and convert it to Workout objects
def import_workout_data(workouts):

    # loop over the workouts in string format and split them by exercise
    for workout in workouts:
        exercises = workout.split('\n')
        workout_list_pre.append(exercises)

    # loop over workouts and their exercises, extract the data and create Set, Exercise and Workout objects that are then added to 'workout_list'
    for workout in workout_list_pre:
        date = None
        exercise_list = []
        for exercise in workout:
            name = ''
            sets = []
            cleaned, notes = separate_notes(exercise)
            if re.search(date_pattern, cleaned) != None:
                date = cleaned
                continue
            for string in cleaned.split(' '):
                if not has_digit(string):
                    name += string + ' '
                if has_digit(string):
                    if '(' in string:
                        temp = string.replace('(', '')
                        temp = temp.replace(')', '')
                        temp = temp.split('x')
                        for i in range(int(temp[0])):
                            sets.append(Set(reps=temp[1], weight=temp[2]))
                        continue
                    if '+' in string:
                        sets.append(get_drop_set(string))
                        continue
                    else:
                        sets.append(get_set(string))
            exercise_list.append(Exercise(name=name.strip(), sets=sets, notes=notes))
        workout_list.append(Workout(exercises=exercise_list, date=date))
    return workout_list

File: components/workout_txt_reader.cpp
#include <regex>
#include <iostream>
#include <string>
#include <fstream>
#include <algorithm>

#include "workout_data_structs.h"
#include "common_types.h"
#include "helpers.h"

std::regex date_pattern("\d{2}\.\d{2}");

StringVector read_and_clean_workout_data(std::string path) {
    std::fstream input_file(path);
    StringVector workouts;
    if (input_file.is_open()) {
        std::string line;
        std::string workout_data;
        while (std::getline(input_file, line)) {
            workout_data += line;
        }
        std::string delim = "\n\n"; 
        workouts = split(workout_data, delim);
        auto it = workouts.begin();
        while (it != workouts.end()) {
            std::string delim = "\n";
            if (!contains_str(*it, delim)) {
                it = workouts.erase(it);
            }
            else {
                it++;
            }
        }
    }
    else {
        std::cerr << "Unable to open file" << std::endl;
    }
    return workouts;
}

Exercise get_exercise_from_str(std::string& string, std::string& notes) {
    std::string name = "";
    char delim = ' ';
    StringVector sets_str = split_string(string, delim);
    std::vector<Set> sets = std::vector<Set>();
    for (auto& s : sets_str) {
        if (!has_digit(s)) {
            name += s + ' '; 
        }
        else {
            char c = '(';
            if (contains_char(s, c)) {
                s.erase(std::remove(s.begin(), s.end(), c), s.end());
                std::string delim = "x";
                StringVector multi_set = split(s, delim);
                for (int i = 0; i < stoi(multi_set[0]); i++) {
                    sets.push_back(Set::normal_set(stoi(multi_set[1]), stod(multi_set[2])));
                }
                continue;
            }
            char plus = '+';
            if (contains_char(s, plus)) {
                sets.push_back(get_drop_set(s));
                continue;
            }
            else {
                sets.push_back(get_set(s));
            }
        }
    }
    Exercise exercise(name, sets, notes);
    return exercise;
}

std::vector<Workout> import_workout_data(StringVector workouts) {
    StringVector exercises;
    std::vector<StringVector> pre_processing_workout_list;
    std::vector<Workout> workout_list;
    for (auto& workout : workouts) {
        char delim = '\n';
        exercises = split_string(workout, delim);
        pre_processing_workout_list.push_back(exercises);
    }

    for (auto& workout : pre_processing_workout_list) {
        std::string date = "";
        std::vector<Exercise> exercise_list = std::vector<Exercise>();
        for (auto& exercise : workout) {
            // std::string name = "";
            std::tuple<std::string, std::string> set_and_notes = separate_notes(exercise);
            std::string cleaned = std::get<0>(set_and_notes);
            std::string notes = std::get<1>(set_and_notes);
            if (std::regex_search(cleaned ,date_pattern)) {
                date = cleaned;
                continue;
            }
            Exercise ex = get_exercise_from_str(cleaned, notes);
            exercise_list.push_back(ex);
        }
        Workout wo(exercise_list, date);// TODO: name workouts based on primary muscle group focus (advanced feature)
        workout_list.push_back(wo);
    }
    return workout_list;
}